cmd/log/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+//! ...
+//! ```
+//!
+//! Use `-m` or `--monitor` to continuously monitor the buffer, otherwise will just print the
+//! current log and exit.
+//!
+//! If an argument is provided, only string buffers that have a name that
+//! contains the argument as a substring, or are in a task that contains
+//! the argument as a substring will be displayed.  
+//!
+//! See the [`ringbuf`
+//! documentation](https://github.com/oxidecomputer/hubris/blob/master/lib/ringbuf/src/lib.rs) for more details.
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Run, Validate};
+use std::collections::BTreeMap;
+use std::iter::zip;
+use std::num::ParseIntError;
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list variables
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short)]
+    monitor: bool,
+    /// how long to delay between polling for new log entries
+    #[clap(long, short, default_value = "100")]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+///
+/// Extract the timestamp from the line
+///
+fn extract_timestamp(l: &str) -> Result<u64, ParseIntError> {
+    let mut line = l.split(':');
+    // the first element before the ':' should be the timestamp
+    let hopefully_number_part = line.next().unwrap();
+    hopefully_number_part.parse()
+}
+
+fn stringbuf_monitor(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    delay: Option<u64>,
+    once: bool,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // There are two possible shapes of ringbufs, depending on the age of the
+        // firmware.
+        // - Raw Ringbuf that is not wrapped by anything.
+        // - Safe Ringbuf that is inside a StaticCell.
+        //
+        // Here we will attempt to handle them both -- first raw, then fallback.
+        let ringbuf_val: Value = Value::Struct(
+            reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+        );
+
+        let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+        Stringbuf::from_value(&cell.cell.value).unwrap()
+    };
+
+    // use a BTree to keep a sorted hashmap of all the log entries,  this prevents reprinting log
```

Comment:
Do we really need to keep the log entries around? This is going to just keep growing over time if humility is run in monitor mode for a long time. Also, there is no use of the old entries once they are printed out to screen. Can we save `last` from current fetch of buffer and compare it to the `last` from next fetch to determine what needs to be printed next time? Yes, we need to be careful about overflows, but I think a combination of previous timestamp and previous last value can be made to work?

Reply:
The comment is a little vague.  We only need the map to keep track of the most recent entries.  The plan is to remove the entries from the map after they are out of date.  This ticket is tracking [that](https://rivosinc.atlassian.net/browse/SW-441).

So using last should be possible, but there is a decent amount of extra logic that using a map hides away.  I did previously do it that way, but this just ended up simpler.

Reply:
It is not clear to me why a btree is necessary here. Maybe I am missing something. But, I was thinking something simple like this should work:

--> Maintain last from previous read as prev_last. Initially, it would be set to 0.
--> Read buffer from device
--> Go from prev_last to current last and print out all entries (taking wrap into account)
--> Update prev_last to current last

Reply:
I rewrote things. to be more along this line.